handler/rfc8628/strategy_hmacsha.go

```diff
@@ -0,0 +1,94 @@
+// Copyright © 2025 Ory Corp
+// SPDX-License-Identifier: Apache-2.0
+
+package rfc8628
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"github.com/ory/x/errorsx"
+
+	"github.com/ory/x/randx"
+
+	"github.com/ory/fosite"
+	enigma "github.com/ory/fosite/token/hmac"
+)
+
+// DefaultDeviceStrategy implements the default device strategy
+type DefaultDeviceStrategy struct {
+	Enigma *enigma.HMACStrategy
+	Config interface {
+		fosite.DeviceProvider
+		fosite.DeviceAndUserCodeLifespanProvider
+		fosite.UserCodeProvider
+	}
+}
+
+var _ RFC8628CodeStrategy = (*DefaultDeviceStrategy)(nil)
+
+// GenerateUserCode generates a user_code
+func (h *DefaultDeviceStrategy) GenerateUserCode(ctx context.Context) (string, string, error) {
+	seq, err := randx.RuneSequence(h.Config.GetUserCodeLength(ctx), h.Config.GetUserCodeSymbols(ctx))
+	if err != nil {
+		return "", "", err
+	}
+	userCode := string(seq)
+	signUserCode, signErr := h.UserCodeSignature(ctx, userCode)
+	if signErr != nil {
+		return "", "", err
+	}
+	return userCode, signUserCode, nil
+}
+
+// UserCodeSignature generates a user_code signature
+func (h *DefaultDeviceStrategy) UserCodeSignature(ctx context.Context, token string) (string, error) {
+	return h.Enigma.GenerateHMACForString(ctx, token)
+}
+
+// ValidateUserCode validates a user_code
+func (h *DefaultDeviceStrategy) ValidateUserCode(ctx context.Context, r fosite.DeviceRequester, code string) error {
+	exp := r.GetSession().GetExpiresAt(fosite.UserCode)
+	if exp.IsZero() && r.GetRequestedAt().Add(h.Config.GetDeviceAndUserCodeLifespan(ctx)).Before(time.Now().UTC()) {
+		return errorsx.WithStack(fosite.ErrDeviceExpiredToken.WithHintf("User code expired at '%s'.", r.GetRequestedAt().Add(h.Config.GetDeviceAndUserCodeLifespan(ctx))))
+	}
+	if !exp.IsZero() && exp.Before(time.Now().UTC()) {
+		return errorsx.WithStack(fosite.ErrDeviceExpiredToken.WithHintf("User code expired at '%s'.", exp))
+	}
+	return nil
```

Comment:
Actually, I think you forgot to validate the code here

Reply:
Good catch, fixed

Reply:
So you store the code plain text in the database? Normally, we only store the HMAC-SHA signature of the token/code in the DB.

Reply:
No, only the signature is stored. 
When we get a user_code, we calculate its signature, fetch the device request from the database and then call `ValidateUserCode`. This is similar to how the `authorization_code` is validated (https://github.com/ory/fosite/blob/master/handler/oauth2/flow_authorize_code_token.go#L30).

The only difference is that the auth_code consists of a plain token + its signature, so you can validate the signature. The `user_code` is just a plain token because it is of limited length, so we can't validate the token on it's own.

Perhaps the name `ValidateUserCode` is a little misleading, as it validates only that the code is still active and not the actual string.

Reply:
Ok, I understand, that makes sense. Maybe we adjust the API naming slightly to make it less confusing, or we add a code comment to this specific part (potentially also just pointing to this comment) to make sure this is preserved as a reason for the future / pen testing.

Reply:
I tried to add a comment about it, hope it makes sense. I decided not to update the method name as I felt that it is better to keep the naming consistent with other strategies.